sqldb/invoices.go

```diff
@@ -0,0 +1,1651 @@
+				InvoiceID: int64(s.invoice.AddIndex),
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+	case invpkg.ContractCanceled:
+		err := s.db.OnInvoiceCanceled(
+			s.ctx, sqlc.OnInvoiceCanceledParams{
+				AddedAt:   s.updateTime.UTC(),
+				InvoiceID: int64(s.invoice.AddIndex),
+			},
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	params := sqlc.UpdateInvoiceStateParams{
+		ID:           int64(s.invoice.AddIndex),
+		State:        int16(newState),
+		SettledIndex: settledIndex,
+		SettledAt:    settledAt,
+	}
+
+	if preimage != nil {
+		params.Preimage = preimage[:]
+	}
+
+	result, err := s.db.UpdateInvoiceState(s.ctx, params)
+	if err != nil {
+		return err
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if rowsAffected == 0 {
+		return invpkg.ErrInvoiceNotFound
+	}
+
+	if settledIndex.Valid {
+		s.invoice.SettleIndex = uint64(settledIndex.Int64)
+		s.invoice.SettleDate = s.updateTime
+	}
+
+	return nil
+}
+
+// UpdateInvoiceAmtPaid updates the invoice amount paid to the new amount.
+func (s *sqlInvoiceUpdater) UpdateInvoiceAmtPaid(
+	amtPaid lnwire.MilliSatoshi) error {
+
+	_, err := s.db.UpdateInvoiceAmountPaid(
+		s.ctx, sqlc.UpdateInvoiceAmountPaidParams{
+			ID:             int64(s.invoice.AddIndex),
+			AmountPaidMsat: int64(amtPaid),
+		},
+	)
+
+	return err
+}
+
+// UpdateAmpState updates the state of the AMP sub invoice identified by the
+// setID.
+func (s *sqlInvoiceUpdater) UpdateAmpState(setID [32]byte,
+	newState invpkg.InvoiceStateAMP, _ models.CircuitKey) error {
+
+	var (
+		settledIndex sql.NullInt64
+		settledAt    sql.NullTime
+	)
+
+	switch newState.State {
+	case invpkg.HtlcStateSettled:
+		nextSettledIndex, err := s.db.NextInvoiceSettledIndex(s.ctx)
+		if err != nil {
+			return err
+		}
+
+		settledIndex = sqlInt64(nextSettledIndex)
+
+		// If the invoice is settled, we'll also update the settle time.
+		settledAt = sql.NullTime{
+			Time:  s.updateTime.UTC(),
+			Valid: true,
+		}
+
+		err = s.db.OnAMPSubInvoiceSettled(
+			s.ctx, sqlc.OnAMPSubInvoiceSettledParams{
+				AddedAt:   s.updateTime.UTC(),
+				InvoiceID: int64(s.invoice.AddIndex),
+				SetID:     setID[:],
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+	case invpkg.HtlcStateCanceled:
+		err := s.db.OnAMPSubInvoiceCanceled(
+			s.ctx, sqlc.OnAMPSubInvoiceCanceledParams{
+				AddedAt:   s.updateTime.UTC(),
+				InvoiceID: int64(s.invoice.AddIndex),
+				SetID:     setID[:],
+			},
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	err := s.db.UpdateAMPSubInvoiceState(
+		s.ctx, sqlc.UpdateAMPSubInvoiceStateParams{
+			SetID:        setID[:],
+			State:        int16(newState.State),
+			SettledIndex: settledIndex,
+			SettledAt:    settledAt,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Commit finalizes the update before it is written to the database. Note that
+// we don't use this directly in the SQL implementation, so the function is just
+// a stub.
```

Comment:
why do we not use this directly in the SQL implementation? Is because all the operations in the sqlInvoicerUpdater are already executed inside a tx that will commit when we are done somewhere else in the code?

Reply:
> Is because all the operations in the sqlInvoicerUpdater are already executed inside a tx that will commit when we are done somewhere else in the code?

Exactly.